### ISSUE TITLE
Seamless looping animations stutter at the last frame before looping

### ISSUE THREAD

**Issue by alfredbaudisch** [godotengine/godot#65513]
### Godot version

4.0 alpha 16 and also 3.5-stable

### System information

Windows 10 Pro 64-bits, RTX 3080 Ti (516.94)

### Issue description

- Godot does not seem able to seamless loop skeletal mesh rotation animations that rotate from 0 to 360 degrees.
- Godot stops for a frame in the last frame before looping.
- To make sure it's not a problem with my animation, I tested the same animation in Blender, Unity, Unreal Engine, Godot 3.5 and Godot 4-alpha16, and only in Godot (both versions) that it doesn't loop perfectly, freezing at the last frame before looping.
- Interpolation is set to Linear (everything as default).

![sample-loopin11](https://user-images.githubusercontent.com/248383/189060795-b8edf755-db55-4d52-85ed-e4d406fac5f7.gif)

- After more investigation, I noticed Godot duplicate the 1st frame when importing animations where the 1st frame is the same as the last frame - this is what causes the apparent stutter:

![image](https://user-images.githubusercontent.com/248383/189208329-776913bc-f00b-4de3-9432-2b68b89080e7.png)

### Animation Setup
1. The 1st setup I tried: animation composed of a bone where the 1st keyframe is at 0º and the last at 354º - this is the one looping seamlessly in the other engines.
2. 2nd setup tried: 1st keyframe at 0º and last at 360º (wrong, because the last frame and 1st frame are at the same position).

### Expected
The rotation animation should loop without pauses.

For example, in UE (in this GIF the animation loops 2 times before the end of the GIF, the freeze at the end is due to the GIF. You can follow the looping by moving your cursor on top of one of the blades and following it - see how it's seamless for 2 revolutions):

![loopingue121](https://user-images.githubusercontent.com/248383/189065767-1913e99f-53df-4379-a613-4636e95a87a7.gif)

Biggest proof that it just works everywhere else. The wheel animation in my game trailer below (starts at 0:11) is EXACTLY the same unaltered in UE, it's the same mesh and Blender NLA action (the difference in the sample project attached below is that I removed the other parts of the mesh from the same .blend file):

https://twitter.com/PardallGames/status/1560658418431434753

### Steps to reproduce

1) In Blender, create an armature, add a bone and create a new animation (Action).

2) Set the bone rotation to Euler Angles:
![image](https://user-images.githubusercontent.com/248383/189061515-7c040849-4a10-44f6-961c-301669559ccd.png)

3) Set the 1st keyframe rotation to Y: 0º and last keyframe (60th frame) to 354º.

4) Assign the bone to any kind of mesh, as long as you can see the mesh rotating.

5) Push Down the Action as a NLA Track and export either as FBX or GLTF.

6) Import the mesh and animation into Godot and play it looping. The last frame freezes for a frame before looping.

### Minimal reproduction project

Godot 4-alpha16 project and Blend file:

[LoopingMachineFreezing.zip](https://github.com/godotengine/godot/files/9524691/LoopingMachineFreezing.zip)

An additional project with a Humanoid with Mixamo animations, which loop correctly in Blender and Unity, but in Godot it has duplicate initial frames:
[Godot4-LoopingHumanoid.zip](https://github.com/godotengine/godot/files/9604591/Godot4-LoopingHumanoid.zip)

Regardless of it being Godot4-alpha16, THE SAME happens in Godot 3.5 - which is the worrisome part.

### Temporary Solutions / Hacks

1. Adjust every animation manually: https://github.com/godotengine/godot/issues/65513#issuecomment-1240907596
2. Use my Godot's editor option "Remove Duplicated First Frames" (no need to change meshes and animations): https://github.com/godotengine/godot/issues/65513#issuecomment-1287828202

**Comment by cybereality** [comment]
I have tested @alfredbaudisch files on my machine (Ubuntu 22.04 and AMD RX 6800 XT) and I see the same thing here. So I don't believe it is OS or GPU/driver related.

**Comment by TokageItLab** [comment]
Let me clear up the issue, this is a loop problem, not a rotation problem, right?

**Comment by alfredbaudisch** [comment]
> Let me clear up the issue, this is a loop problem, not a rotation problem, right?

Yes, correct, a looping problem.

**Comment by alfredbaudisch** [comment]
I managed to make it work, but I'm not sure whether this is just a hack/workaround and whether Godot really has a bug, so I'll keep the issue open and wait for someone to confirm.

TL;DR; the 1st frame of the looping animation should be the second. And the current 1st frame should be moved to frame 0, but not exported.

## Fix
- Move the keyframe that is related to rotation at 0 degrees to keyframe 0 .
- The NLA strip/the animation itself should start at frame 1 (rotation at 5.9º)
- And it should end at whatever frame you want, as long as the last value has the rotation at 360º.
- Now it loops seamless.
- **If the looping animation is not related to rotation, the same logic applies, the first frame should actually be the SECOND.**
- This is not the consistent with any other 3D application or engine, so I guess it's still an issue in Godot?

![image](https://user-images.githubusercontent.com/248383/189167235-1b0186e7-ab37-47e7-90bd-b6359cc28fff.png)

![loopfixed](https://user-images.githubusercontent.com/248383/189168065-2bf786ff-a9bd-447d-be13-dbf432dcb69b.gif)

**Comment by alfredbaudisch** [comment]
I found more details about the bug. Updated the main issue.

**Comment by irmasln** [comment]
I got same problem with 4.0 alpha 17 ,My blender file animation bone have XYZ Eular mode but godot bone transform not work with "XYZ Eular" , so ı had helicopter and engine turning animation not looping.

**Comment by alfredbaudisch** [comment]
@irmasln yes, confirmed, but it's not only rotation. It's any kind of looping.

I have an additional sample project @TokageItLab, this time a humanoid with 4 animations. In all tracks, the 1st frame is duplicated, creating the said stutter. It's specially easy to see in the Walking and Running Casual Animations.

I updated the issue adding a new blend file and reproduction project: `Godot4-LoopingHumanoid.zip`

![image](https://user-images.githubusercontent.com/248383/191179969-568b5bcb-f31b-4d95-b8b9-c00bda250fe7.png)

**Comment by spindlebink** [comment]
I notice this has been marked as "Nice to Fix" on the board, but this seems like a major issue to me--unless you apply a frame offset workaround to every single animation in your game (which for more complex, non-uniformly keyed animations becomes a big deal and which will need to be unapplied when this bug is fixed), this pretty much makes 3D skeletal animations unusable in production.

**Comment by alfredbaudisch** [comment]
> I notice this has been marked as "Nice to Fix" on the board, but this seems like a major issue to me--unless you apply a frame offset workaround to every single animation in your game (which for more complex, non-uniformly keyed animations becomes a big deal and which will need to be unapplied when this bug is fixed), this pretty much makes 3D skeletal animations unusable in production.

Yeah I got quite disappointed seeing "Nice to Fix", it's mandatory - it makes it impossible (or too costly) to use any external or existing animated without redoing all the animations for Godot.

It's worth noting again that things that work consistently in UE, Unity, Blender and Maya do not work in Godot due to this.

**Comment by akien-mga** [comment]
The issues exists both in 3.5 and 4.0 and fixing doesn't require breaking the API - that's what "nice to fix" means - it's not a blocker to get a releasable 4.0 version. It's nice if it can be fixed before 4.0, but it's not dramatic if it ends up fixed in 4.0.2 or 4.1.

**Comment by spindlebink** [comment]
I don't remember this issue (duplicating first frame on import) being present earlier in the 4.0 cycle or in earlier versions of 3.x, but I also did most 4.0 work back before the alphas were a thing. It was immediately noticeable when I got back into using Godot--are we sure this isn't a more recent regression?

**Comment by alfredbaudisch** [comment]
## Hack: change the editor instead of changing every animation
Instead of manually changing EVERY animation from EVERY mesh, I just added a hack into the editor, a new option called **"Remove Duplicated First Frame"** in the Mesh Animation import dialog which fixes this bug. The code is in my fork of Godot: https://github.com/alfredbaudisch/godot/tree/seamless-looping-hack-dev

![GodotTempFix](https://user-images.githubusercontent.com/248383/197347603-0be80846-eef6-498d-a90c-e69ed72771a5.png)

**Comment by alfredbaudisch** [comment]
> @alfredbaudisch I sent a PR, can you check #68665?

@TokageItLab I tested with both of my sample models attached in the issue ([LoopingMachineFreezing.zip](https://github.com/godotengine/godot/files/9524691/LoopingMachineFreezing.zip) and [Godot4-LoopingHumanoid.zip](https://github.com/godotengine/godot/files/9604591/Godot4-LoopingHumanoid.zip)) and with Cueing on the looping animations are now seamless!

**Comment by C0fyC0de** [comment]
I have this problem in good 4.2 but only when using animation tree, any fix?

**Comment by jjbionicle** [comment]
@alfredbaudisch I'm using gotot 4.3 and still getting this issue but your exporting solution isnt working for me. How do I launch your custom version of godot, when I open the folder I don't see an  .exe to launch anywhere.
